src/main/java/org/springframework/integration/kafka/core/DefaultConnectionFactory.java

```diff
@@ -227,14 +162,48 @@ public void refreshLeaders(Collection<String> topics) {
 	 */
 	@Override
 	public Collection<Partition> getPartitions(String topic) {
-		if (!getPartitionBrokerMap().getPartitionsByTopic().containsKey(topic)) {
+		// first, we try to read the topic from the cache. We use the read lock to block if a write is in progress
+		Collection<Partition> returnedPartitions = null;
+		try {
+			lock.readLock().lock();
+			if (getMetadataCache().containsTopic(topic)) {
+				returnedPartitions = getMetadataCache().getPartitions(topic);
+			}
+		}
+		finally {
+			lock.readLock().unlock();
+		}
+		// if we got here, it means that the data was not available, we should try a refresh. The lock is reentrant
+		// so we will not block ourselves
+		if (returnedPartitions == null) {
+			try {
+				lock.writeLock().lock();
+				this.refreshLeaders(Collections.singleton(topic));
+				// if data is not available after refreshing, it means that the topic was not found
+				if (getMetadataCache().containsTopic(topic)) {
+					returnedPartitions = getMetadataCache().getPartitions(topic);
+				}
+			}
+			finally {
+				lock.writeLock().unlock();
+			}
+		}
+		if (returnedPartitions == null) {
 			throw new TopicNotFoundException(topic);
 		}
-		return getPartitionBrokerMap().getPartitionsByTopic().get(topic).toList();
+		return returnedPartitions;
 	}
 
-	private PartitionBrokerMap getPartitionBrokerMap() {
-		return partitionBrokerMapReference.get();
+	private MetadataCache getMetadataCache() {
+		return metadataCacheHolder.get();
+	}
+
+	@SuppressWarnings("serial")
+	private static class ErrorlessTopicMetadataPredicate implements Predicate<TopicMetadata> {
+		@Override
+		public boolean accept(TopicMetadata topicMetadata) {
+			return topicMetadata.errorCode() == ErrorMapping.NoError();
```

Comment:
I don't think that flipping would help, unless we box and use `equals`. But seems very defensive, seeing as `topicMetadata` can't really be `null`, based on its use.